Merge lthaj_WIP to develop

hansmannj opened the pull request:
Hi all

I'd be glad to read your comments if you find some time to look at this.
Unfortunately the "make lint" did not work for all .py files, so I need to sort that out on thursday.
But "make serve" and "make gunicornserve" worked (not for port 8080 though) and the service works as expected and returns colorized symbols.

Thanks a lot
Juergen

hansmannj commented on README.md:
```diff
@@ -78,6 +78,37 @@ None
+
+    "The image to be colorized doesn't exist", 400
+
+Hostname or Domain not part of authorised hosts or domains
```
the check for unauthorized hostnames and domains is not yet implemented

hansmannj replied:
This should now be implemented very similar to service qrcode

hansmannj commented on app/helpers/__init__.py:
```diff
@@ -0,0 +1,32 @@
+from werkzeug.exceptions import BadRequest
+
+
+def check_color_channels(r, g, b):
```
For both, this and the "check_path"-functions, I initially defined return statements consisting of a status_code (e.g. 200 or 400) and a message. These status_codes and response messages were then evaluated within the invoking routes.py and a BadRequest was raised if necessary.

Now the BadRequest will already be raised from within the check-functions, which saves some lines of code and no return statement needs to be defined.

Not sure which way is more elegant, both will do for me.

ltshb replied:
In flask you can use the abort() method with the code and message. This method is then used to raise the correct exception based on the code. Because we use Flask I would use the Flask way to do things and not the Werkzeug.

hansmannj replied:
This is now implemented very similar to service qrcode, using abort()

hansmannj commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
-flask==1.1.1
\ No newline at end of file
+flask==1.1.1
+Pillow==6.2.2
```
Version number might have to be adjusted to >=7.1.0 according to [this alert](https://github.com/geoadmin/mf-chsdi3/network/alert/requirements.txt/Pillow/open)

hansmannj replied:
updated to version 7.2.0

boecklic commented on app/routes.py:
```diff
@@ -0,0 +1,66 @@
+from app.helpers import check_path
+
+# Python2/3 compatibility
+if six.PY3:
+    from io import BytesIO
+else:
+    from StringIO import StringIO
```
We support only py3 for the new services, so all stuff related to py2 should be removed (six et. al)

ltshb replied:
Aggree with this, it make the code less readable.

hansmannj replied:
support of py2 deleted

ltshb commented on app/helpers/__init__.py:
```diff
@@ -0,0 +1,32 @@
+        raise BadRequest('Color channel values must be integers in the range of 0 to 255.')
+
+
+def check_path(path):
+    """
+    This function checks if the specified image file exists.
+    This function will be invoked from routes.py.
+    If everything is well defined, nothing special will happen here.
+    Otherwise a BadRequest will be raised.
+    :param image_name: path of the image file to be checked.
+    :return: nothing.
+    """
+    if not os.path.isfile(path):
+        raise BadRequest('The image to colorize doesn\'t exist')
```
The two lines of code are explicit enough and if they are not reused I would not make a function, it's add only overhead and don't add code readability.

hansmannj replied:
the function is now deleted and implemented in routes.py

ltshb commented on app/routes.py:
```diff
@@ -0,0 +1,66 @@
+import os.path
+# using werkzeug.exceptions in order to get rid of pyramid
+# from pyramid.httpexceptions import HTTPBadRequest and other flask stuff
+# from pyramid.view import view_config
+# from pyramid.response import Response
+from flask import Response
+import six  # for compatibility Python 2 / 3
+from PIL import Image
+from app import app
+from app.helpers import check_color_channels
+from app.helpers import check_path
```
The comment about pyramid is not necessary, it should be removed.
Although import should be grouped as defined in the guideline: https://github.com/geoadmin/doc-guidelines/blob/master/PYTHON.md#5-imports-formatting

hansmannj replied:
Comment removed. Order of imports will be re-checked, once pylint runs without errors

ltshb commented on app/routes.py:
```diff
@@ -0,0 +1,66 @@
+    :param filename: name of the file containing the image/symbol to be colored
+    :return:
+    """
+    check_color_channels(r, g, b)
+    r = int(r)
+    g = int(g)
+    b = int(b)
```
These lines and the check_color_channels() method should be removed and make use of Flask route template type like Chris mentioned.

hansmannj replied:
I rewrote that part

ltshb commented on app/routes.py:
```diff
@@ -0,0 +1,66 @@
+
+    check_path(path)
+
+    mask = Image.open(path)
+
+    if mask.mode == 'P':
+        mask = mask.convert('RGBA')
+    img = Image.composite(Image.new("RGB", mask.size, (r, g, b)), mask, mask)
+    if six.PY3:
+        output = BytesIO()
+    else:
+        output = StringIO()
+    img.save(output, format='PNG')
+
+    return Response(output.getvalue(), mimetype='image/png')
```
The Image Object should be closed, at best it should be used with a 
`with Image.open(path) as img`
See https://pillow.readthedocs.io/en/stable/reference/open_files.html#file-handling

hansmannj replied:
Very good point, thanks.

ltshb commented on app/routes.py:
```diff
@@ -0,0 +1,66 @@
+
+    check_path(path)
+
+    mask = Image.open(path)
```
Why use `mask` ? should not be `img` and below use either `new_img` or `img_converted`

hansmannj replied:
I more or less used the existing code here.
Mask is, because the image is also used as a mask for the composite image.
So for me the naming makes sense, but I am open to rename if required.

ltshb replied:
:+1:

ltshb commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
-flask==1.1.1
\ No newline at end of file
+flask==1.1.1
+Pillow==6.2.2
```
We should use the latest version 7.2.0.

hansmannj replied:
updated to version 7.2.0

ltshb commented:
@hansmannj, @ltkum  and  @boecklic  the route prefix should be `/v4/color`  ? I currently set my prefix in Flask to `/v4/qrcode` I think all three services should be consistent regarding routing prefix.

ltshb commented:
Docu should be in openapi.yaml

ltshb commented:
> Hi all
> 
> I'd be glad to read your comments if you find some time to look at this.
> Unfortunately the "make lint" did not work for all .py files, so I need to sort that out on thursday.
> But "make serve" and "make gunicornserve" worked (not for port 8080 though) and the service works as expected and returns colorized symbols.
> 
> Thanks a lot
> Juergen

Regarding pylint I would first make the required changes and then see if there is still pylint issues. In this case we can check it together to see if a changes in the pylintrc config is required or simply code fix or code ignore.

boecklic commented:
> @hansmannj, @ltkum and @boecklic the route prefix should be `/v4/color` ? I currently set my prefix in Flask to `/v4/qrcode` I think all three services should be consistent regarding routing prefix.

Yeah, go with `/v4/<service-name>` for the moment

boecklic commented:
> See comments from changes.
> 
> Another issue is with all the images assets (*.png) why are they in this repo ? They are apparently not served by the application ! I would avoid as much as possible to push images on a repository, GIT is not the best place to save binary data, this can make the repo growing fast and becoming slow especially if the binary file changes often !

The images are actually served by the service. And those binary assets almost never change, so I think it's ok to have them in git.

hansmannj commented on app/helpers/__init__.py:
```diff
@@ -0,0 +1,32 @@
+    :return: nothing.
+    """
+    if not (r.isdigit() and g.isdigit() and b.isdigit()):
+        raise BadRequest('Color channel values must be integers.')
```
Error messages are now jsonified

hansmannj commented on app/helpers/__init__.py:
```diff
@@ -0,0 +1,32 @@
+        raise BadRequest('Color channel values must be integers in the range of 0 to 255.')
+
+
+def check_path(path):
```
Good point, I deleted the function and implemented the check in the routes.py

hansmannj commented on app/routes.py:
```diff
@@ -0,0 +1,66 @@
+    check_color_channels(r, g, b)
+    r = int(r)
+    g = int(g)
+    b = int(b)
```
Good point, I implemented it in a similar manner now

hansmannj commented:
@boecklic and @ltshb:
Thanks for your comments!
I have addressed your suggested changes and hope everything is OK now.
The openapi.yaml is now up to date, too. Not so familiar with that yet, so I hope, the file is OK.

As you suggested, I implemented the check whether the r, g and b parameters are integers into the route template:
`@app.route('/<int:r>,<int:g>,<int:b>/<string:filename>', methods=['GET'])`.
It works, but I still have to figure out, how to raise a specific error message, in case non-integer values are passed.
This worked before (with the check-functions), but now, simply a 404 with a generic error message will be raised.

Cheers,
Juergen

hansmannj commented on app/routes.py:
```diff
@@ -0,0 +1,66 @@
+
+
+@app.route('/color/<r>,<g>,<b>/<filename>/', methods=['GET'])
+@app.route('/color/<r>,<g>,<b>/<filename>', methods=['GET'])
```
Thanks for that hint!
I implemented that. Now it looks like:
`@app.route('/<int:r>,<int:g>,<int:b>/<string:filename>', methods=['GET'])`

Only problem yet to be resolved is, how to raise a specific error message in case non-integers are passed.
With the check-function used before, it was possible.
Now a generic 404 will be raised.

ltshb commented on app/routes.py:
```diff
@@ -1,66 +1,49 @@
-        output = StringIO()
-    img.save(output, format='PNG')
+        img.save(output, format='PNG')
+        img.close()
```
From my understanding of the pillow api docu the call of close() on an image created by new() or composite() is not necessary.

hansmannj replied:
probably right, but I am not 100% sure. And the close() won't hurt anybody and causes no errors. So I'd opt for keeping it at the Moment

hansmannj commented:
> > @hansmannj, @ltkum and @boecklic the route prefix should be `/v4/color` ? I currently set my prefix in Flask to `/v4/qrcode` I think all three services should be consistent regarding routing prefix.
> 
> Yeah, go with `/v4/<service-name>` for the moment

/v4/ prefix is implemented

boecklic commented on openapi.yaml:
```diff
@@ -1,65 +1,117 @@
+                    message: "An internal error occured"
 
-  /rest/example:
+  /v4/color/{r},{g},{b},{filename}:
```
I think there's a typo, shouldn't it be `/v4/color/{r},{g},{b}/{filename}`?

hansmannj replied:
thanks, I corrected that and also a copy / paste issue (changed qrcode with color a few lines above)

ltshb commented on app/__init__.py:
```diff
@@ -1,11 +1,47 @@
+# Reject request from non allowed origins
+@app.before_request
+def validate_origin():
+    if (
+        'Origin' in request.headers and
+        not re.match(ALLOWED_DOMAINS_PATTERN, request.headers['Origin'])
+    ):
+        abort(make_error_msg(403, 'Not allowed'))
```
After discussion with @boecklic  we decided to always reject request without origin (see https://github.com/geoadmin/service-qrcode/pull/5/commits/4c7cd590c2731020f839df4d86b899eddb031ab2)

hansmannj replied:
this is implemented now

ltshb commented:
> > See comments from changes.
> > Another issue is with all the images assets (*.png) why are they in this repo ? They are apparently not served by the application ! I would avoid as much as possible to push images on a repository, GIT is not the best place to save binary data, this can make the repo growing fast and becoming slow especially if the binary file changes often !
> 
> The images are actually served by the service. And those binary assets almost never change, so I think it's ok to have them in git.

Is the serving of static file already implemented ? I've haven't found anything in code regarding static file serving ?

procrastinatio commented on app/routes.py:
```diff
@@ -0,0 +1,49 @@
+from app.helpers import make_error_msg
+
+# add route prefix
+app.route = prefix_route(app.route, '/v4/color/')
```
Should the versioning not handle by `traefik`?

hansmannj replied:
> Looks very appart the following three points:
> 
> * Origin check in before_request (see related comment)
> * Change the default port from 8080 to 5000 in Makefile and in wsgy.py (Docker should be left to 8080)
> * Logging

first two bullets are implemented
Logging will be done in the next iteration, together with the testing (working on it already).

hansmannj replied:
I'll leave that for discussion. Would be also of relevance for service-qrcode et al.

boecklic replied:
Good point, In case where we really need to operate several versions in parallel it'll either be handled by traefik or directly cloudfront. If it's the same service handling several versions we can think about a pattern like this (dunno if it's possible with prefix_route but you'll get the idea):
```
@app.route('/<string:version>/color/...`)
def foo(version,....):
    if version == 'v1':
        # handle according to v1 
    elif version == 'v2':
        # handle accroding to v2
    else:
        raise UnsupportedVersionError
```
What I would avoid is doing path rewrite in CF and traefik

hansmannj replied:
Simple version handling is now implemented in `app.route = prefix_route(app.route, '/v<int:ver>/color')`. Versions are supposed to be integers, is that OK? And as I do not know about versions 1 to 3, simply all version <= 4 are accepted, and for all versions >4 an "unsupported version" error is raised.

I guess there are more elegant solutions for this, but if it will be handled by traefik or cloudfront, I guess this simple approach might do for now.

ltkum replied:
The way to handle prefixes in flask is with environ['SCRIPT_NAME'] and ['PATH_INFO'] which are actually used within the middleware "reverseproxy". If we want to add a prefix within the code, it should be done here.

If we want to support multiple versions of the same service for some time (I guess this would start with v5), I would not bloat the code with multiple versions and, instead, would deploy two instances (one for each version) and let `traefik` take care of the correspondence between routes and correct service.

That way, we can announce to our users that v4 is deprecated, publish v5, and give them some time (three months seems good to basically change a version number in a route) before decommissioning the v4. This would also avoid maintaining the v4 when the v5/6/7 is out, or mixing versions together within the code.

ltshb replied:
I like the idea of having multiple service version in separate instance and letting traefik handling the routing of the version

hansmannj replied:
OK, so if I understand you (and `middleware.py`) correctly, then this could be the place to change the route and add version info: `app.wsgi_app = ReverseProxy(app.wsgi_app, script_name='/')` inside the apps `__init__.py` by adapting the script_name, right?
But then you and also Marc suggest to to the version handling outside the code handled by `traefik`. Guess we need to make up our "common mind" about this, so we can proceed here and also with the other microservices currently in progress. Maybe we can have a short discussion today during or after the daily?

ltkum replied:
The first part of my message was mostly in "where prefixing should happen when it's within the code", but in our current case, `traefik` can do all the work for us and we won't even need to prefix the routes.

hansmannj commented:
Thanks to all of you. I have tried to implement all of your suggestions.

The next iteration will include:
* logging (already in progress)
* testing (already in progress)
* perhaps a more specific error message when the check in `@app.route('/<int:r>,<int:g>,<int:b>/<string:filename>', methods=['GET'])` fails. Currently only a generic 404 is raised in that case.

A few points left for discussion are:
* should the versioning be handled by traefik instead of in the route definition? This would also be of relevance for service-qrcode (and shortener as well?)
* @ltshb as for the png files in the repo: do you see a more elegant way? 

Appart from that I'd suggest to merge this PR if OK for all of you.
When the issues from the next iterations are ready and tested locally, I'll bug you with a new PR ;-)

boecklic commented on app/routes.py:
```diff
@@ -1,42 +1,56 @@
+    :param ver: integer value of the service's version (inherited from prefix_route).
+    :return: OK with a 200 status code or raise error in case of unsupported version.
     """
+    if ver > 4:
```
I'd make the `4` a variable, we'll probably reset the counting for the internal services to 1

hansmannj replied:
Good point. I added "check_version()" to "check_functions.py" and also defined MAX_ALLOWED_VERSION = 4 in "check_functions.py". In "routes.py" all that is left is "check_version(ver)".
